GopherCI fails with an Internal error when there's missing cgo dependencies

I did a `git push` with 3 new commits to `master` of https://github.com/shurcooL/cmd:

![image](https://cloud.githubusercontent.com/assets/1924134/25553088/97dc9266-2c78-11e7-8cba-c66b6165b17f.png)

And got a 500 from GopherCI:

![image](https://cloud.githubusercontent.com/assets/1924134/25553094/b8a36d94-2c78-11e7-96a8-1a6c94eac48b.png)

The description is "Could not get VCS".

https://gci.gopherci.io/analysis/319

Reporting this in case it's helpful. Feel free to close otherwise.

Thanks @shurcooL 

```
Apr 29 14:37:22 www1 gopherci[2792]: 2017/04/29 14:37:22 queueProcessor: processing error: cannot analyse push event for sha 844accfd52e9b4abe20d858de5490c38ae78293f on repo https://github.com/shurcooL/cmd: could not run analyser: could not execute [install-deps.sh]: [install-deps.sh] returned exit code 2
Apr 29 14:37:22 www1 gopherci[2792]: No (or unsupported) vendoring tool, using go get
Apr 29 14:37:22 www1 gopherci[2792]: # github.com/go-gl/glfw/v3.2/glfw
Apr 29 14:37:22 www1 gopherci[2792]: In file included from ./glfw/src/internal.h:169:0,
Apr 29 14:37:22 www1 gopherci[2792]: from ./glfw/src/context.c:28,
Apr 29 14:37:22 www1 gopherci[2792]: from ../../go-gl/glfw/v3.2/glfw/c_glfw.go:4:
Apr 29 14:37:22 www1 gopherci[2792]: ./glfw/src/x11_platform.h:36:22: fatal error: X11/Xlib.h: No such file or directory
Apr 29 14:37:22 www1 gopherci[2792]: #include <X11/Xlib.h>
Apr 29 14:37:22 www1 gopherci[2792]: ^
Apr 29 14:37:22 www1 gopherci[2792]: compilation terminated.
```

This is because the dependencies cgo needs weren't installed prior. If I install `xorg-dev` and `libgl1-mesa-dev` as defined in the repo's `.travis-ci.yml` install deps works successfully.

Some options, and keen to hear other input

- Detect the use of cgo, don't install deps and mark build as successful with a message `Skipped because cgo is unsupported.`. This would likely be the first step, as the build is currently being marked as failed when it shouldn't.
- Somehow allow the user to define a list of packages to be installed after cloning.

Defining the list of packages to be install could be via:

- `.gopherci.yml` or similar file in the repository. This would be optional and provide per repo customisations. I originally thought this should be a part of a web configuration, but those that have access to configure an installation would only be a small subset of maintainers that would need to install additional packages due to ongoing development. As per #8.
- Detect other CI platform's directives and install those, such as the repo's `.travis-ci.yml` `addons.apt_packages`. This I don't like, it may cause GopherCI to install packages that aren't required at all for GopherCI, but were required for other CI runs.
- Detect a makefile with a target `deps`, and if so, use that to install dependencies. This seems a little inexact, and I think a dotfile would be preferable.

As for detecting cgo (and then gracefully terminating), I'm not sure the best way to do this, are there any easy but correct options?

```
root@a7b6bf3fd525:/go/src/github.com/go-gl/glfw/v3.2/glfw# go list -f '{{ .Imports }}'
[C fmt image image/draw log reflect sync unsafe]
root@a7b6bf3fd525:/go/src/github.com/go-gl/glfw/v3.2/glfw# cd -
/go/src/github.com/shurcooL/cmd
root@a7b6bf3fd525:/go/src/github.com/shurcooL/cmd# go list  -f '{{ join .Deps "\n" }}' ./... | grep C
root@a7b6bf3fd525:/go/src/github.com/shurcooL/cmd# go list  -f '{{ join .CgoFiles "\n" }}' ./...
root@a7b6bf3fd525:/go/src/github.com/shurcooL/cmd# go list -f '{{ .Imports }}' ./... | grep C
root@a7b6bf3fd525:/go/src/github.com/shurcooL/cmd#
```

--- a/integration_test.go
+++ b/integration_test.go
@@ -341,3 +341,16 @@ func TestMissingBeforeRef(t *testing.T) {
 
 	it.WaitForSuccess("master", "ci/gopherci/push")
 }
+
+// TestAPTPackagesCGO tests that cgo dependencies are installed via apt-get.
+func TestAPTPackagesCGO(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+
+	it := NewIntegrationTest(ctx, t)
+	defer it.Close()
+
+	it.Exec("apt-cgo.sh", "aptcgo")
+
+	it.WaitForSuccess("aptcgo", "ci/gopherci/push")
+}

--- a/internal/analyser/analyser.go
+++ b/internal/analyser/analyser.go
@@ -66,7 +66,7 @@ func (e *NonZeroError) Error() string {
 // Analyse downloads a repository set in config in an environment provided by
 // analyser, running the series of tools. Writes results to provided analysis,
 // or an error. The repository is expected to contain at least one Go package.
-func Analyse(ctx context.Context, analyser Analyser, cloner Cloner, tools []db.Tool, config Config, analysis *db.Analysis) error {
+func Analyse(ctx context.Context, analyser Analyser, cloner Cloner, configReader ConfigReader, config Config, analysis *db.Analysis) error {
 	start := time.Now()
 	defer func() {
 		analysis.TotalDuration = db.Duration(time.Since(start))
@@ -88,6 +88,12 @@ func Analyse(ctx context.Context, analyser Analyser, cloner Cloner, tools []db.T
 		return errors.WithMessage(err, "could not clone")
 	}
 	analysis.CloneDuration = db.Duration(time.Since(deltaStart))
+
+	// read repository's configuration
+	repoConfig, err := configReader.Read(ctx, exec)
+	if err != nil {
+		return errors.WithMessage(err, "could not configure repository")
+	}
 
 	// create a unified diff for use by revgrep
 	patch, err := getPatch(ctx, exec, config.BaseRef, config.HeadRef)
@@ -115,7 +121,7 @@ func Analyse(ctx context.Context, analyser Analyser, cloner Cloner, tools []db.T
 	}
 	pwd := string(bytes.TrimSpace(out))
 
-	for _, tool := range tools {
+	for _, tool := range repoConfig.Tools {
 		deltaStart = time.Now()
 		args := []string{tool.Path}
 		for _, arg := range strings.Fields(tool.Args) {

--- a/internal/analyser/analyser.go
+++ b/internal/analyser/analyser.go
@@ -95,6 +95,11 @@ func Analyse(ctx context.Context, analyser Analyser, cloner Cloner, configReader
 		return errors.WithMessage(err, "could not configure repository")
 	}
 
+	// install packages
+	if err := installAPTPackages(ctx, exec, repoConfig.APTPackages); err != nil {
+		return errors.WithMessage(err, "could not install packages")
+	}
+
 	// create a unified diff for use by revgrep
 	patch, err := getPatch(ctx, exec, config.BaseRef, config.HeadRef)
 	if err != nil {
@@ -206,4 +211,16 @@ func getPatch(ctx context.Context, exec Executer, baseRef, headRef string) ([]by
 		}
 	}
 	return patch, nil
+}
+
+// installAptPackages install packages using apt package manager, it expects
+// apt-get update to have already been executed. Can be called with 0 or more
+// packages.
+func installAPTPackages(ctx context.Context, exec Executer, packages []string) error {
+	if len(packages) == 0 {
+		return nil
+	}
+	args := append([]string{"apt-get", "install", "-y"}, packages...)
+	_, err := exec.Execute(ctx, args)
+	return errors.Wrapf(err, "could not install %d apt_packages", len(packages))
 }

--- a/internal/analyser/analyser_test.go
+++ b/internal/analyser/analyser_test.go
@@ -42,16 +42,20 @@ func (c *mockCloner) Clone(context.Context, Executer) error {
 	return nil
 }
 
+type mockConfig struct {
+	RepoConfig RepoConfig
+}
+
+var _ ConfigReader = &mockConfig{}
+
+func (c *mockConfig) Read(context.Context, Executer) (RepoConfig, error) {
+	return c.RepoConfig, nil
+}
+
 func TestAnalyse(t *testing.T) {
 	cfg := Config{
 		BaseRef: "base-branch",
 		HeadRef: "head-branch",
-	}
-
-	tools := []db.Tool{
-		{ID: 1, Name: "Name1", Path: "tool1", Args: "-flag %BASE_BRANCH% ./..."},
-		{ID: 2, Name: "Name2", Path: "tool2"},
-		{ID: 3, Name: "Name2", Path: "tool3"},
 	}
 
 	diff := []byte(`diff --git a/subdir/main.go b/subdir/main.go
@@ -90,8 +94,17 @@ index 0000000..6362395
 	mockDB := db.NewMockDB()
 	analysis, _ := mockDB.StartAnalysis(1, 2)
 	cloner := &mockCloner{}
+	configReader := &mockConfig{
+		RepoConfig{
+			Tools: []db.Tool{
+				{ID: 1, Name: "Name1", Path: "tool1", Args: "-flag %BASE_BRANCH% ./..."},
+				{ID: 2, Name: "Name2", Path: "tool2"},
+				{ID: 3, Name: "Name2", Path: "tool3"},
+			},
+		},
+	}
 
-	err := Analyse(context.Background(), analyser, cloner, tools, cfg, analysis)
+	err := Analyse(context.Background(), analyser, cloner, configReader, cfg, analysis)
 	if err != nil {
 		t.Fatal("unexpected error:", err)
 	}

--- a/internal/analyser/analyser_test.go
+++ b/internal/analyser/analyser_test.go
@@ -68,6 +68,7 @@ index 0000000..6362395
 
 	analyser := &mockAnalyser{
 		ExecuteOut: [][]byte{
+			{},   // installAPTPackages
 			diff, // git diff
 			{},   // install-deps.sh
 			[]byte(`/go/src/gopherci`),                   // pwd
@@ -79,6 +80,7 @@ index 0000000..6362395
 			[]byte("file is generated"),                  // isFileGenerated
 		},
 		ExecuteErr: []error{
+			nil, // installAPTPackages
 			nil, // git diff
 			nil, // install-deps.sh
 			nil, // pwd
@@ -96,6 +98,7 @@ index 0000000..6362395
 	cloner := &mockCloner{}
 	configReader := &mockConfig{
 		RepoConfig{
+			APTPackages: []string{"package1"},
 			Tools: []db.Tool{
 				{ID: 1, Name: "Name1", Path: "tool1", Args: "-flag %BASE_BRANCH% ./..."},
 				{ID: 2, Name: "Name2", Path: "tool2"},
@@ -128,6 +131,7 @@ index 0000000..6362395
 	}
 
 	expectedArgs := [][]string{
+		{"apt-get", "install", "-y", "package1"},
 		{"git", "diff", fmt.Sprintf("%s...%v", cfg.BaseRef, cfg.HeadRef)},
 		{"install-deps.sh"},
 		{"pwd"},
@@ -204,5 +208,38 @@ func TestGetPatch_diffError(t *testing.T) {
 
 	if !reflect.DeepEqual(patch, wantPatch) {
 		t.Errorf("unexpected patch\nhave %v\nwant %v", patch, wantPatch)
+	}
+}
+
+func TestInstallAPTPackages(t *testing.T) {
+	tests := []struct {
+		packages []string
+		args     []string
+	}{
+		{[]string{}, nil},
+		{[]string{"package1"}, []string{"apt-get", "install", "-y", "package1"}},
+		{[]string{"package1", "package2"}, []string{"apt-get", "install", "-y", "package1", "package2"}},
+	}
+
+	for _, test := range tests {
+		analyser := &mockAnalyser{
+			ExecuteOut: [][]byte{{}},
+			ExecuteErr: []error{nil},
+		}
+
+		err := installAPTPackages(context.Background(), analyser, test.packages)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+
+		if test.args == nil {
+			if len(analyser.Executed) > 0 {
+				t.Errorf("expected no execution, have: %v", analyser.Executed)
+			}
+		} else {
+			if !reflect.DeepEqual(test.args, analyser.Executed[0]) {
+				t.Errorf("\nhave: %v\nwant: %v", test.args, analyser.Executed[0])
+			}
+		}
 	}
 }

--- a/internal/analyser/configReader.go
+++ b/internal/analyser/configReader.go
@@ -11,7 +11,8 @@ import (
 
 // RepoConfig contains the analyser configuration for the repository.
 type RepoConfig struct {
-	Tools []db.Tool
+	APTPackages []string `yaml:"apt_packages"`
+	Tools       []db.Tool
 }
 
 // A ConfigReader returns a repository's configuration.

--- a/internal/analyser/configReader.go
+++ b/internal/analyser/configReader.go
@@ -1,0 +1,53 @@
+package analyser
+
+import (
+	"context"
+
+	yaml "gopkg.in/yaml.v1"
+
+	"github.com/bradleyfalzon/gopherci/internal/db"
+	"github.com/pkg/errors"
+)
+
+// RepoConfig contains the analyser configuration for the repository.
+type RepoConfig struct {
+	Tools []db.Tool
+}
+
+// A ConfigReader returns a repository's configuration.
+type ConfigReader interface {
+	Read(context.Context, Executer) (RepoConfig, error)
+}
+
+// YAMLConfig implements a ConfigReader by reading a yaml configuration file
+// from the repositories root.
+type YAMLConfig struct {
+	Tools []db.Tool // Preset tools to use, before per repo config has been applied
+}
+
+var _ ConfigReader = &YAMLConfig{}
+
+// Read implements the ConfigReader interface.
+func (c *YAMLConfig) Read(ctx context.Context, exec Executer) (RepoConfig, error) {
+	cfg := RepoConfig{
+		Tools: c.Tools,
+	}
+
+	const configFilename = ".gopherci.yml"
+
+	args := []string{"cat", configFilename}
+	yml, err := exec.Execute(ctx, args)
+	switch err.(type) {
+	case nil:
+	case *NonZeroError:
+		return cfg, nil
+	default:
+		return cfg, errors.Wrapf(err, "could not read %s", configFilename)
+	}
+
+	if err = yaml.Unmarshal(yml, &cfg); err != nil {
+		return cfg, errors.Wrapf(err, "could not unmarshal %s", configFilename)
+	}
+
+	return cfg, nil
+}

--- a/internal/analyser/configReader_test.go
+++ b/internal/analyser/configReader_test.go
@@ -1,0 +1,80 @@
+package analyser
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+
+	"github.com/bradleyfalzon/gopherci/internal/db"
+)
+
+func TestYAMLConfig_default(t *testing.T) {
+	exec := &mockAnalyser{
+		ExecuteOut: [][]byte{{}},
+		ExecuteErr: []error{&NonZeroError{ExitCode: 1}},
+	}
+
+	reader := &YAMLConfig{
+		Tools: []db.Tool{{Name: "tool1"}},
+	}
+	have, err := reader.Read(context.Background(), exec)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	want := RepoConfig{Tools: reader.Tools}
+
+	if !reflect.DeepEqual(have, want) {
+		t.Errorf("\nhave: %v\nwant: %v", have, want)
+	}
+}
+
+func TestYAMLConfig_unknownError(t *testing.T) {
+	exec := &mockAnalyser{
+		ExecuteOut: [][]byte{{}},
+		ExecuteErr: []error{errors.New("unknown error")},
+	}
+
+	reader := &YAMLConfig{}
+	_, err := reader.Read(context.Background(), exec)
+	if err == nil {
+		t.Errorf("expected error, have: %v", err)
+	}
+}
+
+func TestYAMLConfig_unmarshalError(t *testing.T) {
+	contents := []byte("\t")
+	exec := &mockAnalyser{
+		ExecuteOut: [][]byte{contents},
+		ExecuteErr: []error{nil},
+	}
+
+	reader := &YAMLConfig{}
+	_, err := reader.Read(context.Background(), exec)
+	if err == nil {
+		t.Errorf("expected error, have: %v", err)
+	}
+}
+
+func TestYAMLConfig(t *testing.T) {
+	contents := []byte(`# .gopherci.yml config`)
+	exec := &mockAnalyser{
+		ExecuteOut: [][]byte{contents},
+		ExecuteErr: []error{nil},
+	}
+
+	reader := &YAMLConfig{
+		Tools: []db.Tool{{Name: "tool1"}},
+	}
+	have, err := reader.Read(context.Background(), exec)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	want := RepoConfig{Tools: reader.Tools}
+
+	if !reflect.DeepEqual(have, want) {
+		t.Errorf("\nhave: %v\nwant: %v", have, want)
+	}
+}

--- a/internal/analyser/configReader_test.go
+++ b/internal/analyser/configReader_test.go
@@ -58,7 +58,10 @@ func TestYAMLConfig_unmarshalError(t *testing.T) {
 }
 
 func TestYAMLConfig(t *testing.T) {
-	contents := []byte(`# .gopherci.yml config`)
+	contents := []byte(`# .gopherci.yml config
+apt_packages:
+    - package1
+`)
 	exec := &mockAnalyser{
 		ExecuteOut: [][]byte{contents},
 		ExecuteErr: []error{nil},
@@ -72,7 +75,10 @@ func TestYAMLConfig(t *testing.T) {
 		t.Errorf("unexpected error: %v", err)
 	}
 
-	want := RepoConfig{Tools: reader.Tools}
+	want := RepoConfig{
+		APTPackages: []string{"package1"},
+		Tools:       reader.Tools,
+	}
 
 	if !reflect.DeepEqual(have, want) {
 		t.Errorf("\nhave: %v\nwant: %v", have, want)

--- a/internal/github/handlers.go
+++ b/internal/github/handlers.go
@@ -372,7 +372,11 @@ func (g *GitHub) Analyse(cfg AnalyseConfig) (err error) {
 		GoSrcPath: cfg.goSrcPath,
 	}
 
-	err = analyser.Analyse(ctx, g.analyser, cfg.cloner, tools, acfg, analysis)
+	configReader := &analyser.YAMLConfig{
+		Tools: tools,
+	}
+
+	err = analyser.Analyse(ctx, g.analyser, cfg.cloner, configReader, acfg, analysis)
 	if err != nil {
 		return errors.Wrap(err, "could not run analyser")
 	}

--- a/internal/github/handlers.go
+++ b/internal/github/handlers.go
@@ -162,6 +162,8 @@ func checkPRAction(e *github.PullRequestEvent) error {
 	return nil
 }
 
+const configFilename = ".gopherci.yml"
+
 // checkPRAffectsGo returns true if a pull request modifies, adds or removes
 // Go files, else returns error if an error occurs.
 func checkPRAffectsGo(ctx context.Context, installation *Installation, owner, repo string, number int) (bool, error) {
@@ -172,7 +174,7 @@ func checkPRAffectsGo(ctx context.Context, installation *Installation, owner, re
 			return false, errors.Wrap(err, "could not list files")
 		}
 		for _, file := range files {
-			if hasGoExtension(*file.Filename) {
+			if hasGoExtension(*file.Filename) || *file.Filename == configFilename {
 				return true, nil
 			}
 		}
@@ -188,7 +190,7 @@ func checkPRAffectsGo(ctx context.Context, installation *Installation, owner, re
 func checkPushAffectsGo(event *github.PushEvent) bool {
 	hasGoFile := func(files []string) bool {
 		for _, filename := range files {
-			if hasGoExtension(filename) {
+			if hasGoExtension(filename) || filename == configFilename {
 				return true
 			}
 		}

--- a/internal/github/handlers_test.go
+++ b/internal/github/handlers_test.go
@@ -208,6 +208,10 @@ func TestWebhookHandler(t *testing.T) {
 	// Known good push
 	push := goodPush()
 
+	// Modify .gopherci.yml
+	pushCfg := goodPush()
+	pushCfg.Commits = []github.PushEventCommit{{Modified: []string{configFilename}}}
+
 	// No go files
 	pushNoGo := goodPush()
 	pushNoGo.Commits = []github.PushEventCommit{{Added: []string{"main.php"}}}
@@ -218,6 +222,10 @@ func TestWebhookHandler(t *testing.T) {
 
 	// Known good PR
 	pr := goodPR()
+
+	// Mock API will respond with .gopherci.yml
+	prCfg := goodPR()
+	prCfg.Number = github.Int(3)
 
 	// Mock API will respond with no go files
 	prNoGo := goodPR()
@@ -238,9 +246,11 @@ func TestWebhookHandler(t *testing.T) {
 		wantCode int // http status code we want the response to be
 	}{
 		{push, "push", true, http.StatusOK},
+		{pushCfg, "push", true, http.StatusOK},
 		{pushNoGo, "push", false, http.StatusOK},
 		{pushNoInstall, "push", false, http.StatusOK},
 		{pr, "pull_request", true, http.StatusOK},
+		{prCfg, "pull_request", true, http.StatusOK},
 		{prNoGo, "pull_request", false, http.StatusOK},
 		{prNoInstall, "pull_request", false, http.StatusOK},
 		{prInvalidAction, "pull_request", false, http.StatusOK},
@@ -259,6 +269,10 @@ func TestWebhookHandler(t *testing.T) {
 			fmt.Fprintln(w, "{}")
 		case "/repos/owner/repo/pulls/2/files?per_page=100":
 			file := github.CommitFile{Filename: github.String("main.go")}
+			js, _ := json.Marshal([]*github.CommitFile{&file})
+			fmt.Fprintln(w, string(js))
+		case "/repos/owner/repo/pulls/3/files?per_page=100":
+			file := github.CommitFile{Filename: github.String(configFilename)}
 			js, _ := json.Marshal([]*github.CommitFile{&file})
 			fmt.Fprintln(w, string(js))
 		case "/repos/owner/repo/pulls/4/files?per_page=100":

--- a/testdata/apt-cgo.sh
+++ b/testdata/apt-cgo.sh
@@ -1,0 +1,30 @@
+#!/bin/bash -eux
+
+git checkout -b $1
+
+mkdir cgo
+cat > cgo/main.go <<EOF
+package main
+
+/*
+#include <gsl/gsl_version.h>
+
+const char* ver() {
+        return GSL_VERSION;
+    }
+*/
+import "C"
+
+func main() {
+        println(C.GoString(C.ver()))
+}
+EOF
+
+cat > .gopherci.yml <<EOF
+apt_packages:
+    - libgsl0-dev
+EOF
+
+git add .
+git commit -m "commit"
+git push -f -u origin HEAD

